Descriptive Analytics of US Visa Applications

Context:

A permanent labor certification issued by the Department of Labor (DOL) allows an employer to hire a foreign worker to work permanently in the United States. In most instances, before the U.S. employer can submit an immigration petition to the Department of Homeland Security's U.S. Citizenship and Immigration Services (USCIS), the employer must obtain a certified labor certification application from the DOL's Employment and Training Administration (ETA). The DOL must certify to the USCIS that there are not sufficient U.S. workers able, willing, qualified and available to accept the job opportunity in the area of intended employment and that employment of the foreign worker will not adversely affect the wages and working conditions of similarly employed U.S. workers.

Content:

Data covers 2012-2017 and includes information on employer, position, wage offered, job posting history, employee education and past visa history, associated lawyers, and final decision.

Acknowledgements:

This data was collected and distributed by the US Department of Labor.

To download data, go to this link and click download from data tab. You need to have Kaggle account for this.
https://www.kaggle.com/jboysen/us-perm-visas

In [3]:
import warnings
warnings.simplefilter('ignore')
import pandas as pd
data = pd.read_csv("us_perm_visas.csv")

In [4]:
data.head()

add_these_pw_job_title_9089 agent_city agent_firm_name agent_state  \
0                         NaN        NaN             NaN         NaN   
1                         NaN        NaN             NaN         NaN   
2                         NaN        NaN             NaN         NaN   
3                         NaN        NaN             NaN         NaN   
4                         NaN        NaN             NaN         NaN   

  application_type        case_no case_number case_received_date case_status  \
0             PERM  A-07323-97014         NaN                NaN   Certified   
1             PERM  A-07332-99439         NaN                NaN      Denied   
2             PERM  A-07333-99643         NaN                NaN   Certified   
3             PERM  A-07339-01930         NaN                NaN   Certified   
4             PERM  A-07345-03565         NaN                NaN   Certified   

  class_of_admission              ...               ri_pvt_employment_firm_to  \
0                J-1              ...                                     NaN   
1                B-2              ...                                     NaN   
2               H-1B              ...                                     NaN   
3                B-2              ...                                     NaN   
4                L-1              ...                                     NaN   

  ri_us_workers_considered schd_a_sheepherder     us_economic_sector  \
0                      NaN                NaN                     IT   
1                      NaN                NaN  Other Economic Sector   
2                      NaN                NaN              Aerospace   
3                      NaN                NaN  Other Economic Sector   
4                      NaN                NaN           Advanced Mfg   

  wage_offer_from_9089 wage_offer_to_9089 wage_offer_unit_of_pay_9089  \
0                75629                NaN                          yr   
1                37024                NaN                          yr   
2                47923                NaN                          yr   
3                10.97                NaN                          hr   
4               100000                NaN                          yr   

  wage_offered_from_9089 wage_offered_to_9089  wage_offered_unit_of_pay_9089  
0                    NaN                  NaN                            NaN  
1                    NaN                  NaN                            NaN  
2                    NaN                  NaN                            NaN  
3                    NaN                  NaN                            NaN  
4                    NaN                  NaN                            NaN  

[5 rows x 154 columns]

In [5]:
data.shape

(374362, 154)

In [6]:
for i in data.columns:
    print(i)

add_these_pw_job_title_9089
agent_city
agent_firm_name
agent_state
application_type
case_no
case_number
case_received_date
case_status
class_of_admission
country_of_citizenship
country_of_citzenship
decision_date
employer_address_1
employer_address_2
employer_city
employer_country
employer_decl_info_title
employer_name
employer_num_employees
employer_phone
employer_phone_ext
employer_postal_code
employer_state
employer_yr_estab
foreign_worker_info_alt_edu_experience
foreign_worker_info_birth_country
foreign_worker_info_city
foreign_worker_info_education
foreign_worker_info_education_other
foreign_worker_info_inst
foreign_worker_info_major
foreign_worker_info_postal_code
foreign_worker_info_rel_occup_exp
foreign_worker_info_req_experience
foreign_worker_info_state
foreign_worker_info_training_comp
foreign_worker_ownership_interest
foreign_worker_yr_rel_edu_completed
fw_info_alt_edu_experience
fw_info_birth_country
fw_info_education_other
fw_info_postal_code
fw_info_rel_occup_exp
fw_info

Count of applications submitted by each country

In [7]:
#Here 'country_of_citizenship' is the column with minimum number of null values.
data[['country_of_citizenship','case_number']].groupby(['country_of_citizenship']).count().sort_values("case_number",ascending=False)

case_number
country_of_citizenship               
INDIA                          130707
CHINA                           20177
SOUTH KOREA                     18209
CANADA                           8509
MEXICO                           5282
PHILIPPINES                      4640
UNITED KINGDOM                   2932
TAIWAN                           2487
PAKISTAN                         2309
FRANCE                           2048
VIETNAM                          1996
NEPAL                            1923
BRAZIL                           1900
IRAN                             1811
VENEZUELA                        1762
JAPAN                            1760
RUSSIA                           1553
TURKEY                           1515
GERMANY                          1388
ITALY                            1226
ISRAEL                           1124
AUSTRALIA                        1092
COLOMBIA                         1049
UKRAINE                          1018
SPAIN                             916
POLAND                            909
ARGENTINA                         760
BANGLADESH                        700
NIGERIA                           662
EGYPT                             639
...                               ...
BURUNDI                             2
REPUBLIC OF CONGO                   2
LIBERIA                             2
SOUTH SUDAN                         2
TURKS AND CAICOS ISLANDS            1
SAMOA                               1
LIECHTENSTEIN                       1
ANDORRA                             1
COMOROS                             1
SAO TOME AND PRINCIPE               1
KIRIBATI                            1
CUBA                                1
CAYMAN ISLANDS                      1
NORTH KOREA                         1
CHAD                                1
SINT MAARTEN                        1
YUGOSLAVIA                          1
VANUATU                             1
NETHERLANDS ANTILLES                1
EQUATORIAL GUINEA                   1
SOMALIA                             1
SOVIET UNION                        1
LESOTHO                             1
CENTRAL AFRICAN REPUBLIC            0
ARUBA                               0
MONACO                              0
MOZAMBIQUE                          0
OMAN                                0
PAPUA NEW GUINEA                    0
QATAR                               0

[202 rows x 1 columns]

List top 3 countries with highest number of applications and also show number of applications by them.

In [8]:
#'country_of_citizenship','case_number'
data[['country_of_citizenship','case_number']].groupby(['country_of_citizenship']).count().nlargest(3,columns='case_number')

case_number
country_of_citizenship             
INDIA                        130707
CHINA                         20177
SOUTH KOREA                   18209

Count of applications which were denied

In [9]:
df1=data[(data['case_status']=='Denied')]
df1['case_number'].count()

12781

Count of applicants holding H-1B visas were certified for Permenant Residency (Consider all forms used for writing "H1-B" in data. Consider only H1-B, not H1-B1)

In [10]:
data[(data['class_of_admission']=='H-1B') & (data['case_status']=='Certified')].count()['case_status']

139312

Sorting columns in descending order according to number of missing values in them

In [11]:
data.isnull().sum().sort_values(ascending=False)

orig_file_date                         374081
orig_case_no                           374068
recr_info_job_fair_to                  372569
recr_info_job_fair_from                372564
recr_info_on_campus_recr_to            372497
recr_info_on_campus_recr_from          372490
ji_live_in_dom_svc_contract            371964
foreign_worker_info_education_other    370422
job_info_training_field                369883
job_info_training_num_months           369583
ri_coll_teach_select_date              368632
ri_coll_tch_basic_process              367349
recr_info_coll_teach_comp_proc         367341
job_info_education_other               366593
job_info_alt_combo_ed_other            365148
recr_info_pro_org_advert_to            360292
recr_info_pro_org_advert_from          360286
pw_source_name_other_9089              358603
ri_pvt_employment_firm_to              357922
ri_pvt_employment_firm_from            357920
ri_us_workers_considered               357129
naics_2007_us_title               

Listing top 10 cities from which employers have submitted the most applications. 
How many applications were submitted by them?

In [12]:
data[['employer_city','case_number']].groupby(['employer_city']).count().sort_values("case_number",ascending=False).head(10)


case_number
employer_city               
COLLEGE STATION        11632
NEW YORK                9844
SANTA CLARA             6966
SAN JOSE                6073
MOUNTAIN VIEW           6028
REDMOND                 4901
SAN FRANCISCO           4125
SUNNYVALE               4075
PLANO                   3844
CHICAGO                 3784

Listing top 10 employers who have submitted the highest number of applications

In [13]:
data[['employer_name','case_number']].groupby(['employer_name']).count().sort_values("case_number",ascending=False).head(10)

case_number
employer_name                                             
COGNIZANT TECHNOLOGY SOLUTIONS US CORPORATION        11587
MICROSOFT CORPORATION                                 4625
INTEL CORPORATION                                     4127
GOOGLE INC.                                           3771
AMAZON CORPORATE LLC                                  2716
CISCO SYSTEMS, INC.                                   2097
INFOSYS LTD.                                          1795
APPLE INC.                                            1730
ORACLE AMERICA, INC.                                  1552
FACEBOOK, INC.                                        1268

What is the average time of processing the applications?

In [14]:
data['process_days']=pd.to_datetime(data['decision_date'])-pd.to_datetime(data['case_received_date'])
data['process_days'].mean()

Timedelta('208 days 01:59:13.654466')

What is count of applications for each case status(e.g. Certified, Denied etc) for applicants with "Doctorate" degree?

In [15]:
# YOUR CODE HERE
df1=data[data['foreign_worker_info_education']=='Doctorate']
df1[['case_status','case_number']].groupby(['case_status']).count()

case_number
case_status                   
Certified                 5965
Certified-Expired         5449
Denied                     345
Withdrawn                  484

What are the top 3 economic sectors that has maximum applicants? How many applications are submitted from these sectors? (Use case_no column here, instead of case_number)

In [16]:
data[['us_economic_sector','case_no']].groupby(['us_economic_sector']).count().nlargest(3,columns='case_no')

case_no
us_economic_sector            
IT                       52617
Advanced Mfg             18749
Other Economic Sector    15798

What is the mean wage of all the certified visa applications?

In [17]:
#Identifying the wage column having minimum number of null values among the columns realted to wage
print(data[(['wage_offer_from_9089','wage_offer_to_9089','wage_offer_unit_of_pay_9089','wage_offered_from_9089','wage_offered_to_9089','wage_offered_unit_of_pay_9089'])].isnull().sum().sort_values(ascending=False))
df1=data[data['case_status']=='Certified']
#Please note here we have not considered case number while calculating mean of applicant's wage as case number is relevant
#with respect to visa applications lodged, there will be people with incomplete values (case number) but working in US ,
# therefore in order to cover them we have not considered case number.
df1['wage_offer_from_9089']=pd.to_numeric(df1['wage_offer_from_9089'], errors='coerce')
print( '\nMean wage for wage_offer_from_9089:' )
df1['wage_offer_from_9089'].mean()

wage_offered_to_9089             343347
wage_offered_unit_of_pay_9089    303663
wage_offer_to_9089               300728
wage_offered_from_9089           259715
wage_offer_unit_of_pay_9089      115916
wage_offer_from_9089             114771
dtype: int64

Mean wage for wage_offer_from_9089:


93428.316439185

# Insights

Biggest Employer who applied for Visa application in IT domain was (refer Finding 1(a))
Microsoft=1491(Indian)+535(Chinese)+453(Canadian)=2479,
whereas two Non-American companies, Infosys and HCL are having application number as 915 and 385 respectively. 
Among non-IT companies, biggest applicant is Intel applying for Indians with 1644 application and it is a company 
in Advance Manufacturing.

Whereas if we segregate the applicants on the basis of educational degree(Finding 1(b))
Among top 15 companies, Infosys and HCL are not having any applicant’s with Master’s degree, (Non-American companies),whereas all American companies’  applicants are having Master’s degree .
Companies who have core business in IT (Cognizant and Infosys) domain are having applicants with Bachelor's 
degree in majority.

In [18]:
data1=data
data1['citizenship_country']=data1['country_of_citizenship'].fillna('')+data1['country_of_citzenship'].fillna('')
data1['case_application']=data1['case_no'].fillna('')+data1['case_number'].fillna('')
print('\nFinding 1(a) - Top 15 employers along with economic sector and nationality of applicants ')
print(data1[['employer_name','us_economic_sector','citizenship_country','case_application']].groupby(['employer_name','us_economic_sector','citizenship_country']).count().nlargest(15,columns='case_application'))
print('\nFinding 1(b) - Top 15 employers along with applicants educational degree ')
print(data1[['employer_name','foreign_worker_info_education','case_application']].groupby(['employer_name','foreign_worker_info_education']).count().nlargest(15,columns='case_application'))



Finding 1(a) - Top 15 employers along with economic sector and nationality of applicants 
                                                                                         case_application
employer_name                                 us_economic_sector    citizenship_country                  
INTEL CORPORATION                             Advanced Mfg          INDIA                            1644
MICROSOFT CORPORATION                         IT                    INDIA                            1491
INFOSYS LTD.                                  IT                    INDIA                             915
ORACLE AMERICA, INC.                          IT                    INDIA                             675
AMAZON CORPORATE LLC                          Retail                INDIA                             545
MICROSOFT CORPORATION                         IT                    CHINA                             535
DELOITTE CONSULTING LLP                       Other Economic 

# Report Analysis


a) As per the analysis done by me on the given dataset, it requires lot of data cleaning. 

b)There are multiple columns like case_number and case_no or country_of_citizenship and country_of_citzenship , which shows same category of data.This should be handled by integrating two columns together. 


c)There are lot of columns in the dataset which are having values with more than one data types for e.g wage_offer_from_9089 was having float and string datatypes.We need to handle them properly. 


d)After analysis of the data, we have found that overall nationality wise Indian, Chinese and South Korean employees are the top most applicants.There are overall 139312 applications which have been certified for PR. 


e)Employer wise if we see american companies are more than 85% in visa application process in top 15 list. 


f)College Station is the city with maximum number of visa application whereas Chicago is the city with minimum number of visa applications. 


g)Cognizant is the company with maximum number of visa applications whereas , facebook is the least among top ten companies list. 


h)Average time visa application process takes is 208 days approx.